src/table.rs

```diff
@@ -366,15 +366,26 @@ impl<'s> IntoIterator for &'s Table {
 pub(crate) type KeyValuePairs = IndexMap<InternalString, TableKeyValue>;
 
 fn decorate_table(table: &mut Table) {
-    for (key_decor, value) in table
-        .items
-        .iter_mut()
-        .filter(|&(_, ref kv)| kv.value.is_value())
-        .map(|(_, kv)| (&mut kv.key.decor, kv.value.as_value_mut().unwrap()))
-    {
-        // `key1 = value1`
-        *key_decor = Decor::new(DEFAULT_KEY_DECOR.0, DEFAULT_KEY_DECOR.1);
-        value.decorate(DEFAULT_VALUE_DECOR.0, DEFAULT_VALUE_DECOR.1);
+    for kv in table.items.values_mut() {
+        match &mut kv.value {
+            Item::Value(value) => {
+                kv.key.decor = Decor::new(DEFAULT_KEY_DECOR.0, DEFAULT_KEY_DECOR.1);
+                value.decorate(DEFAULT_VALUE_DECOR.0, DEFAULT_VALUE_DECOR.1);
+            }
+            Item::Table(table) => {
+                kv.key.decor = Decor::new(DEFAULT_KEY_PATH_DECOR.0, DEFAULT_KEY_PATH_DECOR.1);
+                table.decor = Decor::new(DEFAULT_TABLE_DECOR.0, DEFAULT_TABLE_DECOR.1);
+            }
+            Item::ArrayOfTables(array) => {
+                kv.key.decor = Decor::new(DEFAULT_KEY_PATH_DECOR.0, DEFAULT_KEY_PATH_DECOR.1);
+                for table in array.iter_mut() {
+                    table.decor = Decor::new(DEFAULT_TABLE_DECOR.0, DEFAULT_TABLE_DECOR.1);
+                }
+            }
```

Comment:
> Along the way I found a formatting bug for tables, and I ended up fixing that as well.

>`Table::fmt()` wasn't properly applying the right decor to tables and
arrays of tables.

What is being considered a bug in the old version?  Not reformatting the table headers?   I think this is a question of whether you are considering it from the test document perspective or the in-memory data structure perspective.  I was personally viewing this from the document perspective, that we are only formatting the visual table body.  I feel like to consider this from the data structure perspective, we would need to recurse.

Reply:
Yeah, at first it was about reformatting headers (I ran into an issue where  converting an inline table into a standard table would cause the header to have a trailing space). Then I realized that we weren't  formatting sub tables are arrays of tables either.

The invariant I was going far was that calling fmt() on every element of the document would format the whole document. Currently that invariant is mostly upheld except for the cases fixed in this PR.

Reply:
btw I realized there is a bug with `Table::fmt` for the description I gave above, https://github.com/ordian/toml_edit/issues/262

> Yeah, at first it was about reformatting headers (I ran into an issue where converting an inline table into a standard table would cause the header to have a trailing space). Then I realized that we weren't formatting leading and trailing whitespace on sub tables or arrays of tables either.

When converting table types, we should probably reset the formatting to defaults since 99% of the time, the formatting will be counter to the users goal.

And/or we provide a "reset to default" formatting / clear formatting function.

> The invariant I was going for was that calling fmt() on every element of the document would format the whole document. Currently that invariant is mostly upheld except for the cases fixed in this PR.

Thanks for sharing the intent!  That is a worthwhile goal.  I wonder if we should have a separate "format table header".

Either way, maybe we should split this out into a separate PR so we can discuss and iterate on it without being tied up with the rest of the work here?